### DISCUSSION TITLE
set_modelling_rule breaks namespace

### DISCUSSION THREAD

**Discussion by roin93** [FreeOpcUa/opcua-asyncio#516] [Q&A]
Hello together,

I create my server by using a more complex version of the code below.
As soon as I call one of the lines with set_modelling_rule() the namespace is somehow broken.

```
new_node = await target_node.add_variable(node_id, browse_name, [x for x in range(0, 4, 1)], datatype=data_type)
# await new_node.set_modelling_rule("i=83")  # ExposesItsArray
await new_node.set_modelling_rule(True)
```

How I find that out?
I connected to my server via uaExpert. At the left side where I can select which namespace should be highlighted is just empty. 
Did I use it wrong? I wanted to set the node to be "ExposesItsArray"

**Comment by AndreasHeine** [comment]
we dont support that feature!

edit:

yes you need to add_reference() modeling rule is only for mandatory or optional

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/c0ddec78bfed33a21e62c08689c95b64fcf03d6a/asyncua/common/node.py#L654

"""
        Add a modelling rule reference to Node.
        When creating a new object type, its variable and child nodes will not
        be instanciated if they do not have modelling rule
        if mandatory is None, the modelling rule is removed
        """

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/c0ddec78bfed33a21e62c08689c95b64fcf03d6a/asyncua/common/node.py#L634
```python
await node.add_reference(ua.ObjectIds.ModellingRule_ExposesItsArray, ua.ObjectIds.HasModellingRule, True, False)
```
but its just addes the reference! the subnodes will **not** be created and updated automaticaly

**Reply by AndreasHeine** [reply]
"i=83" this is kind of ugly and for humans not readable use the ua.ObjectIds enum for that

**Reply by roin93** [reply]
Do I understand correctly, that I can use the add_reference function to add that reference but I would have trouble by creating the needed subnodes?
I am not sure how I would create such subnodes then. 
I am pretty sure I would need to create references HasComponent but I am not sure how to reference to a part of the node.

Furthermore when I add the reference as posted from you the namespace is broken again. UaExpert says that `Reading the namespace table returned status BadEncodingLimitsExceeded`.

> "i=83" this is kind of ugly and for humans not readable use the ua.ObjectIds enum for that

You are right, that is kind of ugly.

Edit:
Ok, now I am a bit confused. My namespace is still broken even after I removed the set_modelling_rule and the add_reference ... Currently not sure what is now causing the problem.

**Reply by AndreasHeine** [reply]
cloning the addresspace of an other server isnt easy i already told you that there are many pitfalls ;) you need to know the adresspace model concepts of opcua and you need to solve the problems first... there is a reason why no one opensourced an aggregator!
